device/devicegate/filter.go

```diff
@@ -0,0 +1,263 @@
+package devicegate
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/xmidt-org/webpa-common/device"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/xhttp"
+)
+
+const (
+	metadataMapLocation = "metadata_map"
+	claimsLocation      = "claims"
+)
+
+var (
+	emptyVal = struct{}{}
+)
+
+type FilterGateOption func(*FilterGate)
+
+type FilterGate struct {
+	filters        map[string]map[interface{}]struct{}
+	allowedFilters map[string]struct{}
+	lock           sync.RWMutex
+}
+
+type filterRequest struct {
+	Key    string
+	Values []interface{}
+}
+
+func NewFilterGate(options ...FilterGateOption) *FilterGate {
+	fg := &FilterGate{
+		filters:        make(map[string]map[interface{}]struct{}),
+		allowedFilters: make(map[string]struct{}),
+	}
+
+	for _, o := range options {
+		o(fg)
+	}
+
+	return fg
+
+}
+
+func WithAllowedFilters(allowedFilters map[string]struct{}) FilterGateOption {
+	return func(fg *FilterGate) {
+		if allowedFilters != nil {
+			fg.allowedFilters = allowedFilters
+		} else {
+			fg.allowedFilters = make(map[string]struct{})
+		}
+	}
+}
+
+func (f *FilterGate) PrettifyFilters() map[string][]interface{} {
+	copy := make(map[string][]interface{})
+
+	f.lock.RLock()
+	for k, v := range f.filters {
+		var filters []interface{}
+		for filter := range v {
+			filters = append(filters, filter)
+		}
+
+		copy[k] = filters
+	}
+	f.lock.RUnlock()
+
+	return copy
+}
+
+func (f *FilterGate) FiltersCopy() map[string]map[interface{}]struct{} {
```

Comment:
A better design is to make is so that you never need to hand out the filters data structure, and thus never need to copy it.  For example, you could have methods like so:

```
// (consider simplifying so you don't need a map of maps)
func (f *FilterGate) Filters(key string) (map[interface{}]struct{}, bool) {
    f.lock.RLock()
    defer f.lock.RUnlock()
    v, ok := f.filters[key]
    return v, ok
}
```